WSL Support

sh41 opened the pull request:
Hi @joshuataylor, 

This is an experiment in adding support for using IntelliJ IDEA on Windows to develop an Elixir project hosted on WSL. There are a miriad of different combinations of systems that this could run on, so it is just not practicable for me to test them. I'm hoping that some of the friendly community members will step up and give this a spin. I've tested in Windows 11 with Ubuntu 24.04 on the WSL host. Most things seem to work so far, but it definitely feels less snappy than the native linux experience. 

The current PR is the result of multiple iterations over strategies to get this working while also reducing the surface area of code that needed to be changed to accommodate it. I used Jetbrains' AI Assistant extensively, but every line of code was checked by me before committing. 

The final task I asked the AI Assistant to do was summarize this PR, which it has done below. It seems to be a fair summary and covers far more detail than I would have remembered to put in! Also, I cherry picked from #3748 and #3747, so if you don't want either of those fixes included, I can remove them.  

Pre-release build: [intellij-elixir-22.0.2-pre+20251226211412.zip](https://github.com/user-attachments/files/24350970/intellij-elixir-22.0.2-pre%2B20251226211412.zip)


Fixes: #3746 #3470 #2499 #1911 #1384 #3674

AI Generated Summary Follows
===== 

## Summary

This PR adds comprehensive WSL (Windows Subsystem for Linux) support to the IntelliJ Elixir plugin, enabling Windows users to develop Elixir projects hosted on WSL using IntelliJ IDEA on Windows.

### What Works
- ✅ SDK detection and configuration for Elixir/Erlang in WSL (asdf, mise, kerl, kiex)
- ✅ All run configurations (Mix, ExUnit, ESpec, Elixir, IEx)
- ✅ External tools (Credo, Dialyzer, Mix Format)
- ✅ Project creation wizard with WSL SDKs
- ✅ Path conversion (Windows UNC paths ↔ WSL POSIX paths)
- ✅ Graceful BEAM process termination (double SIGINT)

### Known Limitations
- ❌ JPS Builder does not support WSL - use Mix run configurations to compile
- ⚠️ I'm not sure if this has broken Windows, or if Windows is just buggy with Elixir. 
- ⚠️ SDK setup: Click "OK" directly after configuring SDKs (avoid "Apply" then "OK")
- ⚠️ Performance is slower than native Linux

### Testing Environment
- Windows 11 + Ubuntu 24.04 WSL2
- Elixir/Erlang installed via both asdf and mise in WSL

## Issues Addressed

### Intended to resolve:
- Fixes #3746 - MIX_HOME detection for mise/asdf version managers
- Fixes #3470 - SDK path validation for WSL installations
- Fixes #2499 - SDK setup with kerl/kiex in WSL
- Fixes #1911 - asdf SDK detection in WSL
- Fixes #1384 - General WSL support implementation
- Fixes #3674 - SDK dropdown population for WSL/remote projects

### Related Issues (Potential Improvements)
- May help with #3659 - SDK setup issues (improved SDK validation)
- May help with #3716 - SDK not being applied (improved persistence)
- May help with #3715 - Can't add Erlang SDK (better error handling)

## Technical Changes

### Core WSL Infrastructure
- **`WslCompatService`** - Detects WSL paths and converts Windows ↔ POSIX paths
- **`WslSafeCommandLineState`** - Base class for all run configurations with WSL support
- **`DoubleSignalTerminator`** - Graceful BEAM termination (SIGINT×2 instead of SIGKILL)

### SDK Management Improvements
- **Consolidated SDK scanning** - New `SdkHomeScan` utility with WSL path support
- **WSL distribution selection** - Choose WSL distro during SDK setup
- **Fixed persistence issues** - Resolved workspace model mismatches (commits 476893e, f61cd52)
- **Auto-assignment** - Automatically set project SDK when first Elixir SDK is created
- **MIX_HOME detection** - Proper support for mise/asdf version managers (#3746)

### Run Configuration Updates
- **Migrated all run configs** - Elixir, IEx, Mix, ExUnit, ESpec now extend `WslSafeCommandLineState`
- **Test framework refactor** - Extracted common `TestRunnerCommandLineState` base class
- **Process handling** - Replaced `ColoredProcessHandler` with `ElixirProcessHandler`

### Plugin Infrastructure
- **SDK listener timing** - Fixed initialization via startup activity
- **Status bar widget** - Added startup activity + WSL distro display
- **Erlang SDK refresh** - Adjusted order to ensure classpaths are set before Elixir copies
- **Mix import wizard** - Fixed EDT violations and SDK dropdown filtering for WSL

### Testing
- **Comprehensive test coverage** - `ElixirWslSdkTest`, `WslArgumentConversionTest`, `WslExecutableTest`
- **Mock infrastructure** - `MockWslCompatService` for unit tests

### Dependencies & Maintenance
- Updated Kotlin to 2.3.0 (with type inference fix)
- Updated GitHub Actions artifacts (v6/v7)
- README improvements (WSL section, accessibility, linter fixes)
- Added `PluginVersion` utility for canary build versioning

dependabot[bot] committed:
Bump org.gradle.toolchains.foojay-resolver-convention

Bumps org.gradle.toolchains.foojay-resolver-convention from 0.8.0 to 1.0.0.

---
updated-dependencies:
- dependency-name: org.gradle.toolchains.foojay-resolver-convention
  dependency-version: 1.0.0
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

sh41 committed:
♻️ Refactor test framework state classes to reduce duplication
Extracted common test execution logic from ESpec and ExUnit State classes into a shared TestCommandLineState base class.
- Add HasCommandLine interface for configurations that provide commandLine()
- Create generic TestCommandLineStatebase class for test frameworks
- Update ESpec and ExUnit configurations to implement HasCommandLine
- Simplify ESpec and ExUnit State classes to only specify framework name
- Add WSL-compatible process creation using runWithModalProgressBlocking

This reduces code duplication between test framework implementations while maintaining type safety and avoiding impact on configurations that don't need commandLine() (e.g., erl.Configuration).

sh41 committed:
✨ Add `WslCompatService` for WSL path detection and command patching
🤡 Add mockito to mock WSL service

We won't have a WSL instance available on the GitHub Runner, so we need to mock this service.

sh41 committed:
🔥 Remove superfluous plugins from the JPS build

sh41 committed:
✨ WSL-safe process handling with graceful BEAM termination

- `WslSafeCommandLineState` to handle WSL-compatible process creation in one place with modal progress blocking to avoid EDT issues.
- `DoubleSignalTerminator` interface with `ElixirDoubleSignalTerminator` for double SIGINT termination. Uses CTRL-C, wait, CTRL-C rather than a straight up SIGKILL.
- `ElixirProcessHandler` extending `KillableColoredProcessHandler` with graceful shutdown
- Refactor `iex.ProcessHandler` to use `DoubleSignalTerminator` composition

sh41 committed:
♻️ Migrate all run configurations to WslSafeCommandLineState

- All remaining Run Configuration State classes to extend WslSafeCommandLineState & implement HasCommandLine
- Replace ColoredProcessHandler with ElixirProcessHandler
- Rename TestCommandLineState → TestRunnerCommandLineState

All run configurations now use WSL-safe process creation and graceful termination.

sh41 committed:
⚡️ Switch the order of the refresh so that Erlang has the class paths setup when Elixir copies them.

sh41 committed:
✅ Delay SdkTableListener startup until after the ApplicationManager is ready.

SDK table listeners were being set up in static init blocks before ApplicationManager was ready. Move initialization to a ProjectActivity to ensure proper startup order.

This fixes failures observed in `EarlyInitializationTest` where SDK Type classes were attempting to access ApplicationManager.getApplication().messageBus during class loading, before the ApplicationManager was available.

sh41 committed:
♻️ Consolidate SDK Scanning Code

♻ Move all the code for scanning SDK locations in to one place so that WSL checks don't need to be duplicated.
♻️ Refactor `getHomeChooserDescriptor` to use shared implementation in `Type` class.
🔥 Remove unused methods from `sdk/elixir/Type` class
🚨And also apply the linter rules that IntelliJ keeps warning about.

sh41 committed:
✨ WSL Path handling in SDK Discovery

✨ WSL Path on Windows shouldn't add exe/bat
✨ WSL Distro Chooser For First SDK Setup
    - Add user prompt to choose environment (Windows vs. WSL) during SDK creation.
    - Implement WSL-specific path handling for dependency selection.
    - Refactor `fixDependency` and related methods to support WSL-aware SDK home paths.
    - Streamline base path resolution for consistent environment handling.

sh41 committed:
♻️ Fix URL protocol in default documentation links (`http` → `https`).

sh41 committed:
✨ Add WSL Distro to Widget text

sh41 committed:
🐛 Fix SDK validation and persistence timing issues

This fixes workspace model mismatch errors that occurred when creating
Elixir SDKs that depend on newly created Erlang SDKs.

Changes:
- SdkAdditionalData: Pass sdkModel to getErlangSdk() in checkValid() to allow validation against in-memory SDKs not yet persisted to table
- SdkAdditionalData: Make getErlangSdk() read-only to prevent mutations after markAsCommitted(), which violates IntelliJ's immutability contract
- SdkAdditionalData: Store erlangSdkName in constructor for proper persistence without requiring table lookup
- SdkAdditionalData: Simplify writeExternal() to write SDK name directly without validation checks that could block newly created SDKs
- Type.createSdkOfType(): Set additional data immediately when creating Elixir SDK to ensure workspace model entity/bridge synchronization
- AdditionalDataConfigurable: Commit changes immediately after classpath updates to ensure persistence

Error seen was:
"entity was <additional /> while bridge had <additional erlang-sdk-name=...>"
workspace model mismatch error.

sh41 committed:
Maybe Auto-set project SDK when new Elixir SDK is added

Automatically sets the project SDK when a new Elixir SDK is added to the global SDK table in an attempt to reduce friction for new users.

Conditions:
- Only auto-sets if exactly one project is open (avoids ambiguity JDK listener is application wide)
- Only auto-sets if current SDK is null or is an invalid Elixir SDK so as not to overwrite deliberate config
- Preserves non-Elixir SDKs (Java, Python, etc.)

Also includes fixes to SDK reference cleanup methods:
- cleanupOrphanedElixirSdkReferences: Use SDK modificator pattern instead of direct mutation to properly update additional data (see SDKAdditionalData#writeCommit)
- updateElixirSdkReferencesAfterRename: Use SDK modificator pattern for proper write action handling
- configureInternalErlangSdk: Avoid setting additional data if already set during SDK creation (prevents duplicate writes)

sh41 committed:
✨ Refactor WSL support to include all external tools

Extend WSL path conversion beyond run configurations to cover all external tools (Credo, Dialyzer, Mix Format, project wizard).

sh41 committed:
🐛 Set MIX_HOME for mise/asdf setups

MIX_HOME may be set incorrectly if the IDE was started in a location that uses a different MIX_HOME than the selected SDK.

Fixes #3746

sh41 committed:
✏️ Resolve loads of linter warnings in the README.md

sh41 committed:
✏️ Add alt text to all imgs in README
Helps screen readers and stops the inspections from warning about them.

sh41 committed:
✨ cmd line logging

sh41 committed:
♻️ Refactor WSL path conversion to work on `ProcessBuilder` instead of `GeneralCommandLine`

Simplify WSL integration by migrating path and environment variable conversions to `ProcessBuilder`. Ensures that conversion happens "Just in time" and only happens once, removes `RUNS_ON_WSL` metadata requirement.

Also move duplicated implementations into `WslCompatService` rather than repeating them in `WslCompatServiceImpl` and `MockWslCompatService`

sh41 committed:
🐛 Fix SDK configuration persistence and UI feedback in AdditionalDataConfigurable

**Problem:**
- Erlang SDK selection was using UI-only modificator, failing to persist additional data
- Immediate commit in ItemListener was a no-op (EditedSdkModificator.commitChanges() is empty)
- Warning label showed incorrect state by reading from persisted SDK instead of UI
- Unnecessary ItemListener events fired during reset()
- Unsafe null assertions (elixirSdk!!) could cause crashes

**Solution:**
- Use real SDK modificator from elixirSdk.sdkModificator to persist additional data
- Read classpath from UI modificator for immediate warning label updates
- Remove no-op commit call; let Editor.apply() handle persistence (respects Apply/Cancel)
- Extend freeze period to prevent ItemListener firing during initialization
- Replace !! with safe null checks

**Technical details:**
- Add overloaded hasErlangClasspathInRoots() to allow checking of sdkModificator
- Add comments explaining UI vs real modificator distinction
- Import OrderRootType for classpath checking

sh41 committed:
🐛 Fix EDT violation when opening Project Structure from SDK widget

Fixes "Slow operations are prohibited on EDT" error that occurred when clicking "Add Elixir SDK..." from the status bar widget popup. The error was triggered at LocalFileSystemBase.convertToNioFileAndCheck during library type detection when the Project Structure dialog opened.

Changed from the deprecated ActionUtil.invokeAction() to the modern ActionUtil.performAction() API, which properly handles threading for modal dialogs and slow operations.
```
  Stack trace reference:
    at com.intellij.util.SlowOperations.assertSlowOperationsAreAllowed
    at com.intellij.openapi.vfs.impl.local.LocalFileSystemBase.convertToNioFileAndCheck
    ...
    at org.elixir_lang.status_bar_widget.ElixirSdkStatusWidget$createAddSdkAction$1.actionPerformed(ElixirSdkStatusWidget.kt:144)
```

sh41 committed:
🐛 Fix EDT violations in Mix project import wizard

- Wrap Messages.showYesNoCancelDialog() in invokeAndWait() to ensure it runs on EDT
- Wrap deleteIdeaModuleFiles() in invokeAndWait() for safe file operations
- Move scanProjectRoot() slow I/O into Task.Modal with progress indicator

Fixes "Slow operations are prohibited on EDT" errors during Mix project import.

sh41 committed:
♻️ Modernize process handling in Mix import wizard

- Replace deprecated ProcessAdapter with ProcessListener interface.
- Use ProcessHandlerFactory.createColoredProcessHandler() instead of OSProcessHandler to automatically handle ANSI escape codes in progress indicator output via built-in AnsiEscapeDecoder.

sh41 committed:
♻️ Clarify persistence model in SdkAdditionalData

Refactor to make explicit that only the Erlang SDK name is persisted,
with the SDK reference as a transient cache.

- Rename erlangSdk → cachedErlangSdk and mark @Transient
- Fix setErlangSdk() to update both name and cache atomically
- Add type validation (staticIsValidDependency) to all lookups
- Clear cache in readExternal() to force re-resolution after load
- Add getErlangSdkName() for race-free checks in SDK table listeners
- Extract helper methods for lookup logic

Add unit tests for persistence and caching behavior.

sh41 committed:
🐛 Handle Erlang SDK deletion/rename in listeners

Fix race conditions in SDK table listeners when an Erlang SDK is deleted or renamed while Elixir SDKs reference it.

Changes:
- Use getErlangSdkName() instead of getErlangSdk() in listeners to check stored references without triggering auto-discovery
- Show warning notification when Erlang SDK is deleted, listing affected Elixir SDKs that need reconfiguration

This replaces the previous auto-update approach which had issues with IntelliJ's transactional Project Structure dialog.

sh41 committed:
Fix Mix import wizard SDK dropdown on WSL/remote environments

During project import, the wizard uses the default project which has isDefault()=true and projectFilePath=null. This causes ProjectSdksModel to use LocalEelDescriptor, filtering out all WSL/remote SDKs even when importing a WSL/remote project.

Wrap the default project to override isDefault() and projectFilePath, enabling proper EelDescriptor detection based on the import path. This allows configured WSL/remote SDKs to appear in the dropdown immediately, without resorting to auto discovery.

sh41 committed:
🔖 Bump semver patch for canary builds

- Add `PluginVersion` utility to manage version bumping for non-release builds, preventing unnecessary Marketplace update prompts.
- Enhance `buildPlugin` task with messages about artifact timestamp consistency and build location.

sh41 committed:
✨ Add support for detecting Elixir installs via install scripts

- New `mergeElixirInstallScript` to handle `.elixir-install` paths from `elixir-lang.org` installation scripts.
- Integrated this method into SDK scanning for local Windows Local, Windows WSL & Linux

sh41 committed:
✨ Support additional version managers for detecting MIX_HOME directory

- Refactored to generalize detection of version managers with known layouts.
- Added support for `elixir-install` alongside `asdf` and `mise`.

sh41 committed:
⚗️ Try removing autodetection

Autodetection leads to some surprising configurations appearing. Try going without it and just telling the user that the SDK is not valid.

sh41 committed:
🔥 Drop unused Makefile

sh41 committed:
🔧 Exclude buggy kubernetes plugin from the build IDEs

com.intellij.kubernetes logs lots of stacktraces when using the `runIde` task. Removing it makes the logs quieter.

sh41 committed:
🧑‍💻 Import intellij-platform-plugin-template run configs

They seem to have better logging than the defaults.

sh41 committed:
👷 Fail tests if an error is logged.

Add com.intellij.testFramework.TestLoggerFactory.createTestWatcher rule to BasePlatformTestCase to catch and throw errors that are logged during tests.

Prevents error conditions from going unnoticed in the pipeline.

sh41 committed:
👷 Add helper function to capture WARN messages from the logger.

Helps keep test logs quieter and allows assertions on log output.

sh41 committed:
👷 Refactor tests to replace `BasePlatformTestCase` with `PlatformTestCase`.

- Standardized test framework usage across tests, so that log capture is as wide-spread as possible.
- Added `TestLoggerFactory.createTestWatcher` to relevant non-`PlatformTestCase` tests to capture and fail on logged errors.

sh41 committed:
👷 Refactor Erlang code execution in tests to use temporary escripts

- Standardized execution method to avoid command-line parsing issues, especially on Windows.
- Replaced repetitive `Runtime.getRuntime().exec` calls with a reusable `executeErlangCode` helper method.

sh41 committed:
🧪 Skip `PipeLocationTest.testPipesInQuoterTmpDirectory` on Windows due to lack of named pipe support.

sh41 committed:
♻️ refactor(buildSrc): Organize build utilities into domain packages

Move build utilities from flat structure into domain-centric packages:
- versioning/: Plugin versioning logic (PluginVersion, VersionFetcher)
- quoter/tasks/: Quoter-related tasks (StartQuoterTask)

This prepares the codebase for additional platform-specific code and
improves maintainability by grouping related functionality together.

sh41 committed:
✨ Add platform abstraction for Windows dev support

Introduce platform detection and abstraction layer to enable building
on both Windows and POSIX (Linux/macOS) systems.

Key platform differences handled:
- Elixir executables: .bat extension required on Windows
- Mix environment: Windows requires local rebar/hex installation
- Quoter daemon: Windows uses ProcessBuilder, POSIX uses native daemon

Platform-specific implementations (PosixElixirPlatform,
WindowsElixirPlatform, PosixQuoterPlatform, WindowsQuoterPlatform)
isolate OS differences while maintaining identical interfaces.

sh41 committed:
✨ Add ElixirService for platform-aware Elixir management

Introduce ElixirService BuildService to manage Elixir installation and
Mix environment setup with automatic platform detection.

The service provides:
- Platform-specific executable paths (handles .bat on Windows)
- Mix environment variable configuration
- Lazy initialization (builds only when needed)
- Thread-safe singleton pattern for Gradle parallel execution

This encapsulates platform differences that were previously scattered
across build.gradle.kts, making the build script cleaner and more
maintainable.

sh41 committed:
♻️ refactor(build): Refactor QuoterService with platform abstraction

Replace hard-coded POSIX daemon management with platform-aware
implementation that supports both Windows and POSIX.

Changes:
- Use QuoterPlatform interface for platform-specific daemon operations
- Add elixirService parameter (for future use)
- Store process reference for Windows cleanup
- Platform detection and logging for debugging

Windows implementation uses ProcessBuilder to manage the daemon as a
foreground process, while POSIX continues to use the native daemon
command. Both platforms share the same RPC-based status checking.

sh41 committed:
✨ Add platform-aware Gradle tasks for Quoter

Add GetQuoterDepsTask and ReleaseQuoterTask that delegate to
ElixirService for platform-specific Mix command execution.

These tasks replace inline Exec task configuration in build.gradle.kts,
providing:
- CI caching support via declared inputs/outputs
- Developer discoverability via ./gradlew tasks
- Platform-agnostic task definitions

The tasks are thin wrappers that get platform-specific executables and
environment from ElixirService, keeping task code simple.

sh41 committed:
✨ Integrate platform-aware services into build

Update build.gradle.kts to register and use ElixirService and refactored
QuoterService. Tasks now delegate to services for platform-specific
behavior.

Changes:
- Register ElixirService with minimal parameters to avoid circular refs
- Update QuoterService registration to include elixirService dependency
- Replace inline Exec tasks with GetQuoterDepsTask/ReleaseQuoterTask
- Add Erlang availability check with helpful error messages

The build now automatically detects the platform and applies appropriate
behavior without explicit platform checks in the build script.

sh41 committed:
📝 Add Windows development setup instructions

Document prerequisites and troubleshooting for Windows developers:

Prerequisites:
- Erlang/OTP required for running tests (auto-detected via erl command)
- Make required for building Elixir from source
- Installation options via Chocolatey or manual download

Troubleshooting:
- Orphaned Erlang processes
- Elixir build failures
- Path with spaces warnings (cosmetic only)

The build system now fully supports Windows through platform abstraction.
Windows developers can build and test the plugin locally with proper
tooling installed.

sh41 committed:
👷 Enable multi-platform test workflows with OS matrix support

Update shared-test.yml to add support for testing on both Ubuntu and Windows. Adjust matrix configuration, test result names, and artifact uploads to include `matrix.os`.

This is a precursor to adding UI/Integration tests.

sh41 committed:
✅ Reduce 300Kb+ failure messages in `DecompilerTest`

Also resolve linter warnings.

sh41 committed:
🔧 Add .gitattributes to stop line endings being changed.

sh41 committed:
♻️ Refactor `Builder` class to resolve linter warnings

sh41 committed:
🐛 Fix JPS path handling on Windows

In windows, if the working directory was on a different drive to the project directory, the JPS builder would silently replace the drive letter with the wrong one and then silently skip building with Mix.

Switched to using the provided `com.intellij.util.io.URLUtil.extractPath` which handles this.

sh41 committed:
🙈 Update .gitignore to exclude AI-related files (CLAUDE.md, /.claude, /.ai)

sh41 committed:
♻️ Refactor mix task execution into shared MixTaskRunner utility

- Extract MixTaskRunner to eliminate code duplication between wizard and actions
- Add @JvmStatic annotations for Java interop (Root.java)
- Provide void methods for Java/wizard (errors logged) and Result method for Kotlin
- Support null project context (import wizard) and non-null (actions)
- Use modern runWithModalProgressBlocking for project context
- Use legacy Task.Modal for wizard (no project context yet)
- Update Root.java import wizard to use shared runner methods

sh41 committed:
✨ Add automatic Mix dependencies checker with notifications

- Add DepsCheckerStartupActivity to detect outdated deps on project open
- Check if mix.lock is newer than deps/ folder
- Check if deps/ folder is missing or empty
- Show notification with "Install dependencies" action button
- Add Notifier.mixDeps* notification methods
- Add InstallAction for notification-triggered installs
- Register DepsCheckerStartupActivity in plugin.xml

Implements proactive dependency management by alerting users when their Mix dependencies need to be updated.

sh41 committed:
♻️ Refactor project import and configuration logic

- Mark projects as importing to enable Mix compiler settings
- Rename `_buildDir` to `buildDir` for clarity
- Remove unnecessary return in `setProjectRoot`
- Align `DirectoryConfigurator` class usage with convention
- Fix typos in user-facing strings ("Mix projects", "Umbrella App detected")

sh41 committed:
✨ Add Install Mix dependencies to status widget context menu.

sh41 committed:
🚨 Resolve compiler warnings.

sh41 committed:
✨ Add UI testing infrastructure with IntelliJ IDE Starter

- Create testUI source set for UI tests using IntelliJ IDE Starter framework
- Add JUnit 5, Kodein DI, and Kotlinx Coroutines dependencies
- Implement testUI Gradle task with mise SDK integration and proper memory settings
- Add reusable UI helpers: DriverExtensions, IdeTestContext, ProjectOperations
- Configure OS-specific VM options and test environment
- Enable skip searchable options for faster development builds
- Update Kotlin compiler flag from deprecated -Xjvm-default=all to -jvm-default=enable

QuoterService moved to test task only (UI tests don't need quoter dependency)

sh41 committed:
✨ Add Ultimate license activation and project import UI test

- Implement license activation helper for Ultimate edition testing
- Add comprehensive UI test demonstrating:
  - License activation workflow
  - SDK configuration (Erlang and Elixir)
  - Project import with dependency installation
  - Running ExUnit tests via IDE

This test validates the complete Ultimate-specific project setup workflow.

joshuataylor commented:
I'm so sorry @sh41 - in rebasing I've become co-committer to your commits.

Is there an easy way for you to undo this - you should be the committer, as it's also lost the signing.

I'll spinup a Windows 11 install on an old laptop and give this a go!

sh41 commented:
lol, no worries, rebased and all verified again. 

When I pushed it the pipeline failed with something about `org.gradle.toolchains.foojay-resolver-convention` so I cherry picked the update from #3728 to see if it resolves it. 

Earlier in the week pushed some more changes, but never got around to adding a comment about them. I was thinking that it would be good to have some basic UI smoke tests that show that the IDE opens and has basic functionality. Even if we don't run them in the pipeline, it could save dev time confirming that basic functionality isn't broken. My idea was to get something that runs on Linux/Windows/macOs across all the major IDEs. 

As a first step, there's now a new task `testUI` that opens IDEA, activates Ultimate license, sets up Erlang/Elixir, opens the quoter project, installs dependencies and runs the mix tests, it checks that the tool window shows `2 tests passed`. 

This is very experimental, only works in Linux on a machine that already has an Ultimate license and is certainly not something that I belongs in CI yet. 

To get it working I added an action "Install Dependencies" which does `mix local.hex`, `mix local.rebar`, `mix deps.get` with a notification bubble on start up if some trivial checks between `mix.lock` and `/deps` show that an update might be needed. I also added it as an action on the startup widget. It was just easier to do it this way than setup a run configuration for it and I think it's a helpful addition to the plugin. 

I can see that the pipeline is failing again. I'll have a quick look at it, but I'm limited on time today.

sh41 committed:
🔧 Add Maven Central access debugging to shared-test workflow

- Add diagnostic steps to troubleshoot network and dependency resolution issues:
  - DNS/Network/TLS info
  - HTTP HEAD/GET requests to Maven repository
  - Gradle dependency resolution with detailed logs

sh41 commented:
Looks like it's a cloudflare issue. Maybe it'll resolve itself if I leave it a few hours.

sh41 committed:
💄 Add @JvmField annotation so the icon works with plugin.xml

sh41 committed:
📝 Update README with more WSL tips

fixmyworld commented:
How likely is this to ship?

sh41 commented:
This PR was split into #3753 #3754 #3755 #3756 #3757 #3758 & #3759 which have all now been merged, so I'm closing this one.
